miscellaneous maintenance fixes

vchlum opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
* The First bug is regarding creating maintenance while node(s) without ncpus is in PBS. This situation is not usual but it is not a bug to have a node without ncpus. For example, If we add a node that is not up/online, no ncpus is set automatically and the creation of maintenance is denied with error "pbs_rsub: Attribute value error" for any host. This is not correct behavior. IMHO, it is better to simply ignore such a node while creating maintenance.

* The second bug: The maintenance can be created only by a manager or operator but it is possible to delete or alter the maintenance by a regular user.


#### Describe Your Change
* With the first bug, we can simply ignore nodes without ncpus by not throwing an error.

* The second bug needs to test the first char in maintenance id and for such maintenance check the correct permissions. This is done in chk_rescResv_request().


#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
Manual test for first bug:
```
(JESSIE)vchlum@torque3:~$ pbsnodes -v torque4
torque4
     Mom = torque4.ics.muni.cz
     Port = 15002
     pbs_version = unavailable
     ntype = PBS
     state = state-unknown,down
     resources_available.host = torque4
     resources_available.vnode = torque4
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 0
     resources_assigned.vmem = 0kb
     comment = node down: communication closed
     resv_enable = True
     sharing = default_shared
     last_state_change_time = Wed Sep 18 09:12:44 2019

(JESSIE)vchlum@torque3:~$ pbs_rsub -R 1200 -D 3600 --hosts torque4
pbs_rsub: missing host(s)
usage: pbs_rsub [-I seconds] [-m mail_points] [-M mail_list]
                [-N reservation_name] [-u user_list] [-g group_list]
                [-U auth_user_list] [-G auth_group_list] [-H auth_host_list]
                [-R start_time] [-E end_time] [-D duration] [-q destination]
                [-r rrule_expression] [-W otherattributes=value...]
                -l resource_list | --hosts host1 [... hostn]
       pbs_rsub --version
(JESSIE)vchlum@torque3:~$ 
```

PTL maintenance test:
[ptl_maintenance_output.txt](https://github.com/PBSPro/pbspro/files/3625847/ptl_maintenance_output.txt)

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

arungrover commented on src/server/svr_chk_owner.c:
```diff
@@ -550,6 +550,11 @@ chk_rescResv_request(char *resvID, struct batch_request *preq)
 		return NULL;
 	}
 
+	if (resvID[0] == PBS_MNTNC_RESV_ID_CHAR && ! (preq->rq_perm & (ATR_DFLAG_OPWR | ATR_DFLAG_MGWR))) {
```
I do not understand this change. Essentially svr_authorize_resvReq does the same thing right? It checks if the request is coming from an Operator or Manager, if not then it compares the reservation owner with the user requesting a rdel or ralter.

vchlum replied:
@arungrover It is not the same. svr_authorize_resvReq() authorizes all Operators and Managers for all reservation, but this change rejects access for all regular users accessing maintenance reservation. It fixes the following problem: A user with the Manager role creates maintenance. Later, the Manager role is removed for this user but the user is still able to remove the maintenance.

vchlum commented:
> @vchlum Thanks for making this contribution.
> We have seen a few minor bugs in maintenance reservation implementation. I do not intend to derail this contribution, it is definitely useful for all of us. But, if you are interested, following are few issues:
> 
> * When a Maintenance reservation is submitted, it can be submitted with -I  option but this option is only useful because the server needs to report the response of scheduler to the user. Since in case of maintenance reservation requests never reach scheduler this option does not really go with "--hosts" option.
> * If "--hosts" option is given with wrong syntax like a comma-separated list of hosts then it considers the last-mentioned hostname as the host where the reservation will run.
> * When a maintenance reservation is confirmed on the nodes where there is already a running reservation then the nodes are taken away from that running reservation and it gets a new "IN-CONFLICT" substate. But when the maintenance reservation ends, the "IN-CONFLICT" reservation never gets the node back again. I am not sure if this is by design or really an anomaly.

Thank you @arungrover for sharing issues.
* You are right. The -l parameter should be forbidden with --hosts. I will fix it in this contribution.
* If you use --hosts option, a space-separated list of hosts is expected. If some host is not found in pbs, it is simply ignored. It means that if you have: '--hosts host1,host2 host3' then 'host1,host2' is considered to be one host. This host is not found and is ignored. The second host is 'host3'. This host is recognized and the maintenance is confirmed on this host only. How do you suggest to change the behavior? I see a possible change like this: in case of an error in hosts syntax, we could always return an error (and do not confirm any maintenance)?
* I consider this to be by design. Do we want to change this behavior? or simply add the fact into the design document?

vchlum commented:
I added the fix for the first issue that @arungrover mentioned. The -l and --hosts can not be used together now.
[ptl_maintenance_output2.txt](https://github.com/PBSPro/pbspro/files/3634677/ptl_maintenance_output2.txt)

arungrover commented on src/cmds/pbs_rsub.c:
```diff
@@ -143,6 +144,7 @@ process_opts(int argc, char **argv, struct attrl **attrp, char *dest)
 				break;
 
 			case 'l':
+				opt_res_req_flg = TRUE;
```
@vchlum my comment was about option -I (ATTR_inter) but this works too because --hosts will consume all ncpus on the hosts anyways. Can you please add similar change to "ATTR_inter" option as well?

arungrover commented:
> * If you use --hosts option, a space-separated list of hosts is expected. If some host is not found in pbs, it is simply ignored. It means that if you have: '--hosts host1,host2 host3' then 'host1,host2' is considered to be one host. This host is not found and is ignored. The second host is 'host3'. This host is recognized and the maintenance is confirmed on this host only. How do you suggest to change the behavior? I see a possible change like this: in case of an error in hosts syntax, we could always return an error (and do not confirm any maintenance)?
> * I consider this to be by design. Do we want to change this behavior? or simply add the fact into the design document?

For the first problem, I feel we should not confirm the reservation if the host does not exist. It gives no indication to the user that reservation has been accepted for only those nodes which are visible to PBS server. 
One more problem with not checking the hosts is that if the same host is given twice in pbs_rsub command, it confirms the reservation on the host. This has no impact as such but it does make server leak a small amount of memory. Here is trace you get from this leak - 
`
==82992== 32 (16 direct, 16 indirect) bytes in 1 blocks are definitely lost in loss record 61 of 1,524

==82992==    at 0x4C29BE3: malloc (vg_replace_malloc.c:299)
==82992==    by 0x44CD00: set_nodes (in /opt/pbs/sbin/pbs_server.bin)
==82992==    by 0x46E940: assign_resv_resc (in /opt/pbs/sbin/pbs_server.bin)
==82992==    by 0x46EBDF: req_confirmresv (in /opt/pbs/sbin/pbs_server.bin)
==82992==    by 0x4B070B: process_socket (in /opt/pbs/sbin/pbs_server.bin)
==82992==    by 0x4B08C1: wait_request (in /opt/pbs/sbin/pbs_server.bin)
==82992==    by 0x42AB89: main (in /opt/pbs/sbin/pbs_server.bin)`

Regarding the second problem, my opinion is that at least the reservation which is marked in conflict should be treated as degraded and another node should be found to keep it running. Currently, a node vanishes underneath the reservation and it never gets it back. The design document does state that IN_CONFLICT state is treated as DEGRADED and gets reconfirmed on other nodes.
Maybe we can have a forum discussion about it and see what others think.
Thanks!

vchlum commented:
I addressed your comments @arungrover. The branch is in conflict now and AppVeyor will not run. Should I rebase?

I will revive the maintenance topic in the forum regarding the last issue.

vchlum commented:
New PTL log: [ptl_maintenance_output3.txt](https://github.com/PBSPro/pbspro/files/3645922/ptl_maintenance_output3.txt)

vchlum commented:
@arungrover It is rebased. There is a problem with Travis on deb 9. I tried several times with the same error. Seems like other PR have the same issue...

arungrover commented:
I have restarted Travis build again. Now that PTL problem is fixed, it might pass.
@prakashcv13 I don't know if you reviewed recent commits, can you please sign-off again if you are ok with the change?

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -792,6 +816,9 @@ main(int argc, char *argv[], char *envp[])
 			int host_ncpus = 0;
 
+			if (strlen(*hostp) == 0)
+				continue;
```
@vchlum, I am not sure, if we really need this check.

vchlum replied:
It handles the special case 'pbs_rsub ... --hosts ""'. It can happen e.g. within the PTL test. I suggest to handle it instead of changing the test.

prakashcv13 replied:
Ok. Then, I believe we should handle this in process_opts()?

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -863,6 +890,13 @@ main(int argc, char *argv[], char *envp[])
 			} /* end of part that crafts select */
+
+			/* host not found or host has zero ncpus */
+			if (host_ncpus == 0) {
```
this if belongs just above the block where we craft the select statement.

prakashcv13 commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -68,9 +68,9 @@ def test_maintenance_acl_denied(self):
         self.assertEqual("pbs_rsub: Unauthorized Request", msg)
 
-    def test_maintenance_forbidden_resources(self):
+    def test_maintenance_forbidden_parameters(self):
```
shall we say "conflicting" parameters?

vchlum commented:
> A few minor comments.

Thank you @prakashcv13. Please, see the latest changes.

prakashcv13 commented:
@arungrover - now it is your turn to review the latest changes :). if you find them good, please go ahead and merge this PR. Thank you.

vchlum commented:
@arungrover It is rebased.
